packages/ckeditor5-ui/src/autocomplete/autocompleteview.ts

```diff
@@ -0,0 +1,232 @@
+				resultsView.element!.scrollTop = 0;
+			} else if ( config.resetOnBlur ) {
+				this.queryView.reset();
+			}
+		} );
+
+		// Update the visibility of the results view when the user types in the query field.
+		// This is an integration for `queryMinChars` configuration.
+		// This is an integration for search results changing length and the #resultsView requiring to be repositioned.
+		this.on( 'search', () => {
+			this._updateResultsVisibility();
+			this._updateResultsViewWidthAndPosition();
+		} );
+
+		// Hide the results view when the user presses the ESC key.
+		this.keystrokes.set( 'esc', ( evt, cancel ) => {
+			resultsView.isVisible = false;
+			cancel();
+		} );
+
+		// Update the position of the results view when the user scrolls the page.
+		// TODO: This needs to be debounced down the road.
+		this.listenTo( global.document, 'scroll', () => {
+			this._updateResultsViewWidthAndPosition();
+		} );
+
+		// Hide the results when the component becomes disabled.
+		this.on( 'change:isEnabled', () => {
+			this._updateResultsVisibility();
+		} );
+
+		// Update the value of the query field when the user selects a result.
+		this.filteredView.on<FilteredViewExecuteEvent>( 'execute', ( evt, { value } ) => {
+			// Focus the query view first to avoid losing the focus.
+			this.focus();
+
+			// Resetting the view will ensure that the #queryView will update its empty state correctly.
+			// This prevents bugs related to dynamic labels or auto-grow when re-setting the same value
+			// to #queryView.fieldView.value (which does not trigger empty state change) to an
+			// #queryView.fieldView.element that has been changed by the user.
+			this.reset();
+
+			// Update the value of the query field.
+			this.queryView.fieldView.value = this.queryView.fieldView.element!.value = value;
+
+			// Finally, hide the results view. The focus has been moved earlier so this is safe.
+			resultsView.isVisible = false;
+		} );
+
+		// Update the position and width of the results view when it becomes visible.
+		resultsView.on( 'change:isVisible', () => {
+			this._updateResultsViewWidthAndPosition();
+		} );
+	}
+
+	/**
+	 * Updates the position of the results view on demand.
+	 */
+	private _updateResultsViewWidthAndPosition() {
+		const resultsView = ( this.resultsView as AutocompleteResultsView );
+
+		if ( !resultsView.isVisible ) {
+			return;
+		}
+
+		resultsView._width = new Rect( this.queryView.fieldView.element! ).width;
+
+		const optimalResultsPosition = AutocompleteView._getOptimalPosition( {
+			element: this.resultsView.element!,
+			target: this.queryView.element!,
+			fitInViewport: true,
+			positions: AutocompleteView.defaultResultsPositions
+		} );
+
+		// _getOptimalPosition will return null if there is no optimal position found (e.g. target is off the viewport).
+		resultsView._position = optimalResultsPosition ? optimalResultsPosition.name as string : 's';
```

Comment:
I know that `AutocompleteResultsView._position` is never `undefined` in your code. But if it would be typed as such, you wouldn't need to cast to `string` (nor use `!`) here and everything would compile anyway.